### DISCUSSION TITLE
cors issue with nextjs

### DISCUSSION THREAD

**Discussion by youssef2891** [vercel/community#704] [Help]
I have a nextjs app that working fine with no issues in development with `npm run dev `and `npm run start` . 
however when I deployed to vercel I get CORS erreurs 
`has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource`

although I am not using next internal API ( dont have an Api folder on my project). all my backend is in a different server `myBackendUrl.com` but someone with similar issue in stackoverflow suggested to add this to `next.config.js `: 
`  async rewrites() {
    return [
      {
        source: "/api/:path*",
        destination: "https://myBackendUrl.com/:path*"
      }
    ]
  },`

also found another answer of someone suggesting to add this to `next.config.js `:
```
  async headers() {
    return [
      {
        source: "/(.*)",
        headers: [
          { key: "Access-Control-Allow-Credentials", value: "true" },
          { key: "Access-Control-Allow-Origin", value: "*" },
          { key: "Access-Control-Allow-Methods", value: "GET,OPTIONS,PATCH,DELETE,POST,PUT" },
          { key: "Access-Control-Allow-Headers", value: "X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version" }
        ]
      }
    ]
  },
```

I did try all these configuration but no success , note also that my backend use to serve a React App (before migrating to Nextjs) and that all cors configuration were set , and that my React App used to work with no issues .

I was able to partially fix this issue when I added to my backend CORS config : `preflightContinue : True'

Any Help Please

**Comment by amyegan** [comment]
> although I am not using next internal API ( dont have an Api folder on my project)

I'm wondering whether the headers settings or proxy would work if you add an internal API layer and call through it to reach the backend. But I want to make sure I understand your setup before I start making wild suggestions. Can you tell me a little more about it or share the repo (or a minimal reproduction) just so I can see what's happening?

**Comment by m-avagyan** [comment]
I configured my next.js middleware to handle accepted origins and return website or route only if origin type is accepted for production mode. For local development I set origin value as * because no need to handle CORS for localhost. You can copy paste this code to your middleware.ts file.

```typescript
import { NextResponse } from 'next/server';
import type { NextRequest } from 'next/server';

const isProduction = process.env.DATASET === 'production';
const corsOrigins = isProduction
    ? [
          'domain1',
          'domain2',
          'domain3',
          'domain4',
      ]
    : ['*'];

export function middleware(req: NextRequest) {
    const origin = req.headers.get('origin');
    const res = NextResponse.next();

    // Set CORS headers
    res.headers.set('Access-Control-Allow-Credentials', 'true');
    res.headers.set('Access-Control-Allow-Methods', 'GET,DELETE,PATCH,POST,PUT');
    res.headers.set(
        'Access-Control-Allow-Headers',
        'X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version'
    );

    if (corsOrigins.includes('*') || corsOrigins.includes(origin || '')) {
        res.headers.set('Access-Control-Allow-Origin', origin || '*');
    }

    return res;
}

export const config = {
    matcher: '/api/:path*', // Match all API routes
};
```

Please modify the corsOrigins array value (add all your needed domains).
